include/seqan3/range/view/kmer_hash.hpp

```diff
@@ -12,47 +12,163 @@
+    
+    //!\brief Construct from a view and a given shape.
+    kmer_hash_view(urng_t urange_, shape const s_) : urange{std::move(urange_)}, shape_(s_)
+    {}
+        
+    //!\brief Construct from a non-view that can be view-wrapped and a given shape.
+    template <std::ranges::ViewableRange orng_t>
+    kmer_hash_view(orng_t && urange_, shape const s_) : 
+        urange{std::view::all(std::forward<orng_t>(urange_))}, shape_(s_)
+    {}
+    //!\}
+
+    /*!\name Iterators
+     * \{
+     */
+    /*!\brief Returns an iterator to the first element of the range.
+     * \returns Iterator to the first element.
+     *
+     * ### Complexity
+     *
+     * Constant.
+     *
+     * ### Exceptions
+     *
+     * No-throw guarantee.
+     */
+    auto begin() noexcept
+    {
+        return iterator{std::ranges::begin(urange), shape_};
+    }
+    
+    //!\copydoc begin()
+    auto begin() const noexcept
+    {
+        return const_iterator{std::ranges::begin(urange), shape_};
+    }
+    
+    //!\copydoc begin()
+    auto cbegin() const noexcept
+    {
+        return const_iterator{std::ranges::begin(urange), shape_};
+    }
+    
+    /*!\brief Returns an iterator to the element following the last element of the range.
+     * \returns Iterator to the end.
+     *
+     * \details
+     *
+     * This element acts as a placeholder; attempting to dereference it results in undefined behaviour.
+     *
+     * ### Complexity
+     *
+     * Constant.
+     *
+     * ### Exceptions
+     *
+     * No-throw guarantee.
+     */
+    auto end() noexcept
+    {
+        return std::ranges::end(urange);
+    }
+    
+    //!\copydoc end()
+    auto end() const noexcept
+    {
+        return std::ranges::end(urange);
+    }
+    
+    //!\copydoc end()
+    auto cend() const noexcept
+    {
+        return std::ranges::end(urange);
+    }
+    //!\}
+};
+
+//!\brief A deduction guide for the view class template.
+template <std::ranges::ViewableRange orng_t>
+kmer_hash_view(orng_t &&, shape const shape_) -> kmer_hash_view<std::ranges::all_view<orng_t>>;
+
+// ---------------------------------------------------------------------------------------------------------------------
+// kmer_hash_fn (adaptor definition)
+// ---------------------------------------------------------------------------------------------------------------------
 
 //![adaptor_def]
 //!\brief view::kmer_hash's range adaptor object type (non-closure).
 struct kmer_hash_fn
 {
-    //!\brief Store the argument and return a range adaptor closure object.
-    constexpr auto operator()(size_t const k) const noexcept
+    //!\brief Returns an instance of seqan3::detail::kmer_hash_view constructed with \p urange and \p shape_.
+    template <std::ranges::ForwardRange urng_t>
+    auto operator()(urng_t && urange, shape const shape_) const
     {
-        return detail::adaptor_from_functor{*this, k};
```

Comment:
static assert on ViewableRange and ForwardRange and Semialphabet inside. See for example view::translate

Reply:
If Semialphabet is required?